python/semantic_kernel/orchestration/sk_function.py

```diff
@@ -189,11 +221,28 @@ async def _local_stream_func(client, request_settings, context):
                     # list of <role, content> messages)
                     completion = ""
                     messages = await chat_prompt.render_messages_async(context)
-                    async for partial_content in client.complete_chat_stream_async(
-                        messages, request_settings
+
+                    # With data case - stream and get the tool message for citations
+                    if request_settings.data_source_settings is not None and hasattr(
+                        client, "complete_chat_stream_with_data_async"
                     ):
-                        completion += partial_content
-                        yield partial_content
+                        response = await client.complete_chat_stream_with_data_async(
+                            messages, request_settings
+                        )
+                        async for partial_content in response:
+                            completion += partial_content
+                            yield partial_content
```

Comment:
this will loop through the content, but then you will not have the sources in the front-end until the full response is in, are the sources also streamed?

Reply:
Moved the tool message processing before the loop!